algorithms/src/sorting/impl/Kokkos_SortByKeyImpl.hpp

```diff
@@ -0,0 +1,322 @@
+void sort_by_key_cudathrust(
+    const Kokkos::Cuda& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    MaybeComparator&&... maybeComparator) {
+  const auto policy = thrust::cuda::par.on(exec.cuda_stream());
+  auto keys_first   = ::Kokkos::Experimental::begin(keys);
+  auto keys_last    = ::Kokkos::Experimental::end(keys);
+  auto values_first = ::Kokkos::Experimental::begin(values);
+  thrust::sort_by_key(policy, keys_first, keys_last, values_first,
+                      std::forward<MaybeComparator>(maybeComparator)...);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ROCTHRUST)
+template <class KeysDataType, class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties, class... MaybeComparator>
+void sort_by_key_hipthrust(
+    const Kokkos::HIP& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    MaybeComparator&&... maybeComparator) {
+  const auto policy = thrust::cuda::par.on(exec.cuda_stream());
+  auto keys_first   = ::Kokkos::Experimental::begin(keys);
+  auto keys_last    = ::Kokkos::Experimental::end(keys);
+  auto values_first = ::Kokkos::Experimental::begin(values);
+  thrust::sort_by_key(policy, keys_first, keys_last, values_first,
+                      std::forward<MaybeComparator>(maybeComparator)...);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ONEDPL) && \
+    (ONEDPL_VERSION_MAJOR > 2022 ||  \
+     (ONEDPL_VERSION_MAJOR == 2022 && ONEDPL_VERSION_MINOR >= 2))
+template <class KeysDataType, class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties, class... MaybeComparator>
+void sort_by_key_onedpl(
+    const Kokkos::SYCL& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    MaybeComparator&&... maybeComparator) {
+  if (keys.stride(0) != 1 && values.stride(0) != 1) {
+    Kokkos::abort(
+        "SYCL sort_by_key only supports rank-1 Views with stride(0) = 1.");
+  }
+
+  // Can't use Experimental::begin/end here since the oneDPL then assumes that
+  // the data is on the host.
+  auto queue  = exec.sycl_queue();
+  auto policy = oneapi::dpl::execution::make_device_policy(queue);
+  const int n = keys.extent(0);
+  oneapi::dpl::sort_by_key(policy, keys.data(), keys.data() + n, values.data(),
+                           std::forward<MaybeComparator>(maybeComparator)...);
+}
+#endif
+
+template <typename ExecutionSpace, typename PermutationView, typename ViewType>
+void applyPermutation(ExecutionSpace const& space,
+                      PermutationView const& permutation,
+                      ViewType const& view) {
+  static_assert(std::is_integral<typename PermutationView::value_type>::value);
+
+  auto view_copy = Kokkos::create_mirror(space, view);
+  Kokkos::deep_copy(space, view_copy, view);
```

Comment:
Any good reason not to do
```suggestion
  auto view_copy = Kokkos::create_mirror(Kokkos::view_alloc{space, typename ExecutionSpace::memory_space{}, Kokkos::WithoutInitializing}, view);
  Kokkos::deep_copy(space, view_copy, view);
```
?